src/huggingface_hub/hf_file_system.py

```diff
@@ -667,6 +730,30 @@ def _upload_chunk(self, final: bool = False) -> None:
                 path=self.resolved_path.unresolve(),
             )
 
+    def read(self, length=-1):
+        """Read remote file.
+
+        If `length` is not provided or is -1, the entire file is downloaded and read. On POSIX systems and if
+        `hf_transfer` is not enabled, the file is loaded in memory directly. Otherwise, the file is downloaded to a
+        temporary file and read from there.
+        """
+        if self.mode == "rb" and (length is None or length == -1):
+            # Open a temporary file to store the downloaded content
+            if HF_HUB_ENABLE_HF_TRANSFER:
+                with tempfile.TemporaryDirectory() as tmp_dir:
+                    # if hf_transfer, we want to provide a real temporary file so hf_transfer can write concurrently to it
+                    tmp_path = os.path.join(tmp_dir, "tmp_file")
+                    self.fs.get_file(self.resolved_path.unresolve(), tmp_path)
+                    with open(tmp_path, "rb") as f:
+                        return f.read()
+            else:
+                # otherwise, we don't care where the file is stored (e.g. in memory or on disk)
+                with tempfile.TemporaryFile() as tmp_file:
+                    self.fs.get_file(self.resolved_path.unresolve(), tmp_file)
+                    return tmp_file.read()
```

Comment:
I'm not a big fan of this code, so maybe `hf_transfer` should support writing the fetched bytes to in-memory buffers

Reply:
AFAIK it's not possible to run hf_transfer in-memory. The current implementation here is not the best but still works and is purely based on public methods.

Reply:
> AFAIK it's not possible to run hf_transfer in-memory

Indeed, but let's then open an issue in the `hf_transfer` repo? Besides the use case here, `hf_transfer` is meant for power users, and they have a lot of RAM (we can assume), so supporting in-memory downloads makes sense from that point of view, too.

Reply:
@mariosasko @lhoestq I'll make a release of `hfh` soon (today?) with this PR. We can discuss/implement an in-memory download for `hf_transfer` but that will not be ready in time. So for this release, would you prefer:
1.  to have `hf_transfer` download to tmp file, then load it in memory and delete tmp file (current implementation). Pro: allows for `hf_transfer` to work. Con: more I/O operations.
2. to disable `hf_transfer` when using `fs.read()`. Pro: no extra I/O ops. Cons: no hf_transfer.
3. wait for an hf_transfer PR before merging and releasing this one. Pro: no in-between solution. Cons: will delay the `get_file` fix.

Reply:
First, note that `.read()` can be called multiple times, e.g.

```
header = f.read(5)
full_data = header + f.read()
```

Then I think it's ok to not have the full speed for `.read()` - big files are generally downloaded to disk and `.read()` is generally used for smaller files (readme, json, images, audio...). I would also expect it to work in memory.

Finally `fsspec` fiels have the notion of `block_size` and cache, which you are bypassing in the current implementation.

I think it's fine to not override `.read()` for now. Especially since the original issue is about fast download to disk

Reply:
If you really want to ship a speed improvement in `.read()` for today's release, maybe you can check if `self.loc == 0` and use `HfFileSystemStreamFile.read()` ?

Reply:
Updated in https://github.com/huggingface/huggingface_hub/pull/2143/commits/ae9e2f70c3b796346ce1e6f7b8739500239173dc.